pkg/controller/uniteddeployment/adapter/statefulset_adapter.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	set.Namespace = ud.Namespace
+
+	if set.Labels == nil {
+		set.Labels = map[string]string{}
+	}
+	for k, v := range ud.Spec.Template.StatefulSetTemplate.Labels {
+		set.Labels[k] = v
+	}
+	for k, v := range ud.Spec.Selector.MatchLabels {
+		set.Labels[k] = v
+	}
+	set.Labels[alpha1.ControllerRevisionHashLabelKey] = revision
+	// record the subset name as a label
+	set.Labels[alpha1.SubSetNameLabelKey] = subsetName
+
+	if set.Annotations == nil {
+		set.Annotations = map[string]string{}
+	}
+	for k, v := range ud.Spec.Template.StatefulSetTemplate.Annotations {
+		set.Annotations[k] = v
+	}
+
+	set.GenerateName = getSubsetPrefix(ud.Name, subsetName)
+
+	selectors := ud.Spec.Selector.DeepCopy()
+	selectors.MatchLabels[alpha1.SubSetNameLabelKey] = subsetName
+
+	if err := controllerutil.SetControllerReference(ud, set, a.Scheme); err != nil {
+		return err
+	}
+
+	set.Spec.Selector = selectors
+	set.Spec.Replicas = &replicas
+	if ud.Spec.Template.StatefulSetTemplate.Spec.UpdateStrategy.Type == appsv1.OnDeleteStatefulSetStrategyType {
+		set.Spec.UpdateStrategy.Type = appsv1.OnDeleteStatefulSetStrategyType
+	} else {
+		if set.Spec.UpdateStrategy.RollingUpdate == nil {
+			set.Spec.UpdateStrategy.RollingUpdate = &appsv1.RollingUpdateStatefulSetStrategy{}
+		}
+		set.Spec.UpdateStrategy.RollingUpdate.Partition = &partition
+	}
+
+	set.Spec.Template = *ud.Spec.Template.StatefulSetTemplate.Spec.Template.DeepCopy()
+	if set.Spec.Template.Labels == nil {
+		set.Spec.Template.Labels = map[string]string{}
+	}
+	set.Spec.Template.Labels[alpha1.SubSetNameLabelKey] = subsetName
+	set.Spec.Template.Labels[alpha1.ControllerRevisionHashLabelKey] = revision
+
+	set.Spec.RevisionHistoryLimit = ud.Spec.Template.StatefulSetTemplate.Spec.RevisionHistoryLimit
+	set.Spec.PodManagementPolicy = ud.Spec.Template.StatefulSetTemplate.Spec.PodManagementPolicy
+	set.Spec.ServiceName = ud.Spec.Template.StatefulSetTemplate.Spec.ServiceName
+	set.Spec.VolumeClaimTemplates = ud.Spec.Template.StatefulSetTemplate.Spec.VolumeClaimTemplates
+
+	attachNodeAffinity(&set.Spec.Template.Spec, subSetConfig)
+	attachTolerations(&set.Spec.Template.Spec, subSetConfig)
+
+	return nil
+}
+
+// PostUpdate does some works after subset updated. StatefulSet will implement this method to clean stuck pods.
+func (a *StatefulSetAdapter) PostUpdate(ud *alpha1.UnitedDeployment, obj runtime.Object, revision string, partition int32) error {
+	set := obj.(*appsv1.StatefulSet)
+	if set.Spec.UpdateStrategy.Type == appsv1.OnDeleteStatefulSetStrategyType {
+		return nil
+	}
+
+	// If RollingUpdate, work around for issue https://github.com/kubernetes/kubernetes/issues/67250
+	return a.deleteStuckPods(set, revision, partition)
+}
+
+// IsExpected checks the subset is the expected revision or not.
+// The revision label can tell the current subset revision.
+func (a *StatefulSetAdapter) IsExpected(ud *alpha1.UnitedDeployment, obj metav1.Object, revision string) bool {
```

Comment:
ud is not used.

Reply:
Used in `uniteddeployment_update.go` L53.